enhancements/oc/image-content-source-policy-awareness.md

```diff
@@ -0,0 +1,87 @@
+  - "@soltysh"
+  - "@wking"
+approvers:
+  - "@smarterclayton"
+creation-date: 2020-05-19
+last-updated: 2020-05-19
+status: provisional
+---
+
+# ImageContentSourcePolicy (ICSP) Awareness
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  
+
+There have been several bugs opened around the experience of a 
+user in a disconnected environment using `oc adm release` commands.  If
+using a mirrored image and the mirrored source registry is connected, 
+the following commands do not succeed when in a disconnected environment:
+
+```console
+$ oc adm release extract --tools aprivateregistry/repo/name:tag
+$ oc adm release mirror aprivateregistry/repo/name:tag --to someregistry/repo/name
+$ oc adm must-gather
+```
+
+This is because the mirrored image tags (the individual component images from a payload)
+retain references to the mirrored registry, usually something like 
+`quay.io/openshift-release-dev/ocp-v4.0-art-dev`.  
+
+There needs to be logic in `oc` to look for `ImageContentSourcePolicy` either from a file or from a cluster.
+`oc` should look for `ICSP` in the cluster/current context if connected, and if not, from a file in an expected location,
+probably from the current directory.  `oc` should gather information about RepositoryDigestMirrors from ICSP and use that
+when extracting or mirroring images, or when running must-gather.
+
+Current bugs regarding this Issue:   
+* https://bugzilla.redhat.com/show_bug.cgi?id=1823839
+* https://bugzilla.redhat.com/show_bug.cgi?id=1823143 and also for 4.3, 4.5, 4.6
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+* In what order should `oc` gather info about an image?  Need to order the following:
+    * ICSP from file
+    * ICSP from cluster
+    * image reference from user-given image
+
+* Should oc adm release mirror write an ICSP file to the current directory rather than a print to stdout?
+* Should there be a flag to direct oc to re-order the above 3 choices, to tell oc which information source to use?
+
+## Summary
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  
+
+## Proposal
+
+* Add logic to `oc adm release` and/or openshift/library-go/pkg/image to become aware of ICSP file
+* Add logic to `oc adm release` and/or openshift/library-go/pkg/image to become aware of ICSP in cluster
+* Add logic to `oc adm release` and/or openshift/library-go/pkg/image to use the ICSP to complete extracts, mirroring, must-gather
+
+### User Stories
```

Comment:
Hey, @sallyom, could this enhancement be a work around for this other [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1728135)  as well?
If that's so, could you take it into account?

Thanks!

Reply:
yes, thanks, i'll incorporate debug as well

Reply:
for oc debug, in disconnected, is it expected that you'd pass `--image myrepo/tools:mytag` to debug command?

Reply:
debug has similar flow like must-gather see my earlier comment.